Add Volume Storage Adapter for Devfiles

maysunfaisal opened the pull request:
Signed-off-by: Maysun J Faisal <maysun.j.faisal@ibm.com>

**What type of PR is this?**
/kind feature

**What does does this PR do / why we need it**:
This PR implements the Storage Adapter for Devfiles. It reads the parsed Devfile data, creates the storage and adds them to the component deployment.

For Acceptance Criteria, pls have a look at the issue post https://github.com/openshift/odo/issues/2472#issuecomment-579966277

#### List of changes:
- Reads the Devfile Aliased Components and checks if they have Volumes
- Creates a PVC with the Devfile volume name if it does not exist, else reuses the PVC volume
- Different Devfile components mentioning the same volume name, shares the volume(why? pls see issue #2510 and https://github.com/openshift/odo/issues/2472#issuecomment-579966277 for discussion)
- Iterates through the list of PVC for a Devfile and adds
-- PVC to the pod template spec
-- Volume Mounts to the pod template spec's containers
-- Adds a default storage of `1Gi`, see issue https://github.com/openshift/odo/issues/2511 on how this can be improved
- tests

**Which issue(s) this PR fixes**:

Fixes #2472

**How to test changes / Special notes to the reviewer**:

##### All pkg adapter tests (includes new `TestStorageAdapter`, `TestStorageCreate` and updated `TestComponentAdapter` ):
```
maysuns-mbp:adapters maysun$ pwd
/Users/maysun/go/src/github.com/openshift/odo/pkg/devfile/adapters
maysuns-mbp:adapters maysun$ go clean -testcache && go test ./... -v
=== RUN   TestNewPlatformAdapter
=== RUN   TestNewPlatformAdapter/get_platform_adapter
--- PASS: TestNewPlatformAdapter (0.00s)
    --- PASS: TestNewPlatformAdapter/get_platform_adapter (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/devfile/adapters	0.592s
?   	github.com/openshift/odo/pkg/devfile/adapters/common	[no test files]
?   	github.com/openshift/odo/pkg/devfile/adapters/kubernetes	[no test files]
=== RUN   TestComponentAdapter
=== RUN   TestComponentAdapter/Case:_Invalid_devfile
=== RUN   TestComponentAdapter/Case:_Valid_devfile
 •  Waiting for component to start  ...
 ✓  Waiting for component to start [128563ns]
--- PASS: TestComponentAdapter (0.00s)
    --- PASS: TestComponentAdapter/Case:_Invalid_devfile (0.00s)
    --- PASS: TestComponentAdapter/Case:_Valid_devfile (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/devfile/adapters/kubernetes/component	0.765s
=== RUN   TestStorageAdapter
=== RUN   TestStorageAdapter/Case:_Invalid_devfile
=== RUN   TestStorageAdapter/Case:_Valid_devfile
--- PASS: TestStorageAdapter (0.00s)
    --- PASS: TestStorageAdapter/Case:_Invalid_devfile (0.00s)
    --- PASS: TestStorageAdapter/Case:_Valid_devfile (0.00s)
=== RUN   TestStorageCreate
=== RUN   TestStorageCreate/storage_test
--- PASS: TestStorageCreate (0.00s)
    --- PASS: TestStorageCreate/storage_test (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/devfile/adapters/kubernetes/storage	0.430s
?   	github.com/openshift/odo/pkg/devfile/adapters/kubernetes/utils	[no test files]
```

##### All the kclient pkg tests:
```
maysuns-mbp:kclient maysun$ pwd
/Users/maysun/go/src/github.com/openshift/odo/pkg/kclient
maysuns-mbp:kclient maysun$ go clean -testcache && go test ./... -v
=== RUN   TestCreateDeployment
=== RUN   TestCreateDeployment/Case:_Valid_deployment_name
=== RUN   TestCreateDeployment/Case:_Invalid_deployment_name
--- PASS: TestCreateDeployment (0.00s)
    --- PASS: TestCreateDeployment/Case:_Valid_deployment_name (0.00s)
    --- PASS: TestCreateDeployment/Case:_Invalid_deployment_name (0.00s)
=== RUN   TestUpdateDeployment
=== RUN   TestUpdateDeployment/Case:_Valid_deployment_name
=== RUN   TestUpdateDeployment/Case:_Invalid_deployment_name
--- PASS: TestUpdateDeployment (0.00s)
    --- PASS: TestUpdateDeployment/Case:_Valid_deployment_name (0.00s)
    --- PASS: TestUpdateDeployment/Case:_Invalid_deployment_name (0.00s)
=== RUN   TestGenerateContainer
=== RUN   TestGenerateContainer/#00
=== RUN   TestGenerateContainer/container1
--- PASS: TestGenerateContainer (0.00s)
    --- PASS: TestGenerateContainer/#00 (0.00s)
    --- PASS: TestGenerateContainer/container1 (0.00s)
=== RUN   TestGeneratePodTemplateSpec
=== RUN   TestGeneratePodTemplateSpec/podSpecTest
--- PASS: TestGeneratePodTemplateSpec (0.00s)
    --- PASS: TestGeneratePodTemplateSpec/podSpecTest (0.00s)
=== RUN   TestGeneratePVCSpec
=== RUN   TestGeneratePVCSpec/1
=== RUN   TestGeneratePVCSpec/2
--- PASS: TestGeneratePVCSpec (0.00s)
    --- PASS: TestGeneratePVCSpec/1 (0.00s)
    --- PASS: TestGeneratePVCSpec/2 (0.00s)
=== RUN   TestWaitAndGetPod
=== RUN   TestWaitAndGetPod/phase:_running
 •  Waiting for component to start  ...
 ✓  Waiting for component to start [181543ns]
=== RUN   TestWaitAndGetPod/phase:_failed
 •  Waiting for component to start  ...
 ✗  Waiting for component to start [71530ns]
=== RUN   TestWaitAndGetPod/phase:_unknown
 •  Waiting for component to start  ...
 ✗  Waiting for component to start [60171ns]
--- PASS: TestWaitAndGetPod (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_running (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_failed (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_unknown (0.00s)
=== RUN   TestCreatePVC
=== RUN   TestCreatePVC/Case:_Valid_pvc_name
=== RUN   TestCreatePVC/Case:_Invalid_pvc_name
--- PASS: TestCreatePVC (0.00s)
    --- PASS: TestCreatePVC/Case:_Valid_pvc_name (0.00s)
    --- PASS: TestCreatePVC/Case:_Invalid_pvc_name (0.00s)
=== RUN   TestAddPVCToPodTemplateSpec
=== RUN   TestAddPVCToPodTemplateSpec/podSpecTest
--- PASS: TestAddPVCToPodTemplateSpec (0.00s)
    --- PASS: TestAddPVCToPodTemplateSpec/podSpecTest (0.00s)
=== RUN   TestAddVolumeMountToPodTemplateSpec
=== RUN   TestAddVolumeMountToPodTemplateSpec/podSpecTest
=== RUN   TestAddVolumeMountToPodTemplateSpec/podSpecTest#01
--- PASS: TestAddVolumeMountToPodTemplateSpec (0.00s)
    --- PASS: TestAddVolumeMountToPodTemplateSpec/podSpecTest (0.00s)
    --- PASS: TestAddVolumeMountToPodTemplateSpec/podSpecTest#01 (0.00s)
=== RUN   TestGetPVCsFromSelector
=== RUN   TestGetPVCsFromSelector/Case:_Valid_pvc_name
=== RUN   TestGetPVCsFromSelector/Case:_Wrong_Label_Selector
--- PASS: TestGetPVCsFromSelector (0.00s)
    --- PASS: TestGetPVCsFromSelector/Case:_Valid_pvc_name (0.00s)
    --- PASS: TestGetPVCsFromSelector/Case:_Wrong_Label_Selector (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/kclient	0.418s
```


##### To test it out manually:
- Set experiment to true `odo preference set experimental true`
- Copy devfile to project dir and execute `odo push` (Pls remember to check if your devfile has volumes) for example:

```
components:
  -
    type: dockerimage
    image: websphere-liberty:19.0.0.3-webProfile7
    alias: runtime
    volumes:
      - name: m3
        containerPath: /home/user/.m3runtime
      - name: m2
        containerPath: /home/user/.m2runtime
      - name: m2
        containerPath: /home/user/.m2runtime222
  -
    type: dockerimage
    alias: tools
    image: quay.io/eclipse/che-java8-maven:nightly
    volumes:
      - name: m2
        containerPath: /home/user/.m2tools
      - name: m3
        containerPath: /home/user/.m3tools
      - name: m4
        containerPath: /home/user/.m4tools
```
- Execute `kubectl get po,deploy,pvc` and confirm the pvc created
- Execute `kubectl describe po <component pod>` and confirm the volume mounts for the containers

```
maysuns-mbp:scripts maysun$ oc get pvc
NAME                                   STATUS   VOLUME            CAPACITY   ACCESS MODES   STORAGECLASS   AGE   LABELS
persistentvolumeclaim/m2-spring-gbrz   Bound    che-workspace7    8Gi        RWO,RWX                       2h    component=spring,storage-name=m2
persistentvolumeclaim/m3-spring-vijx   Bound    che-workspace8    8Gi        RWO,RWX                       16s   component=spring,storage-name=m3
persistentvolumeclaim/m4-spring-hnie   Bound    che-workspace6    8Gi        RWO,RWX                       16s   component=spring,storage-name=m4
```
```
maysuns-mbp:.odo maysun$ oc get po
NAME                                READY   STATUS    RESTARTS   AGE
spring-b5dcd448b-fjf6f              2/2     Running   0          41m

maysuns-mbp:.odo maysun$ oc describe po spring-b5dcd448b-fjf6f 
Name:               spring-b5dcd448b-fjf6f
[...]
Containers:
  runtime:
    Container ID:   docker://d232eb20c57c0550a403a6e58c790a6b9322227b9c78446556eb8f8a9c3a2843
    Image:          websphere-liberty:19.0.0.3-webProfile7
    [...]
    Mounts:
      /home/user/.m2runtime from m2-spring-gbrz-kcmxb-volume (rw)
      /home/user/.m2runtime222 from m2-spring-gbrz-kcmxb-volume (rw)
      /home/user/.m3runtime from m3-spring-vijx-mwljq-volume (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-k6wc8 (ro)
  tools:
    Container ID:   docker://7b976f1b0d8a5935defb6ebb9411432fc5ceed1c5e06df63c37de98ccc2fc564
    Image:          quay.io/eclipse/che-java8-maven:nightly
    [...]
    Mounts:
      /home/user/.m2tools from m2-spring-gbrz-kcmxb-volume (rw)
      /home/user/.m3tools from m3-spring-vijx-mwljq-volume (rw)
      /home/user/.m4tools from m4-spring-hnie-yxpxy-volume (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-k6wc8 (ro)
[...]
Volumes:
  m3-spring-vijx-mwljq-volume:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  m3-spring-vijx
    ReadOnly:   false
  m2-spring-gbrz-kcmxb-volume:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  m2-spring-gbrz
    ReadOnly:   false
  m4-spring-hnie-yxpxy-volume:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  m4-spring-hnie
    ReadOnly:   false
  default-token-k6wc8:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-k6wc8
    Optional:    false
[...]
Events:
  Type    Reason     Age   From                         Message
  ----    ------     ----  ----                         -------
  Normal  Scheduled  32s   default-scheduler            Successfully assigned che/spring-b5dcd448b-fjf6f to maysun-okd-worker3
  Normal  Pulling    30s   kubelet, maysun-okd-worker3  pulling image "websphere-liberty:19.0.0.3-webProfile7"
  Normal  Pulled     29s   kubelet, maysun-okd-worker3  Successfully pulled image "websphere-liberty:19.0.0.3-webProfile7"
  Normal  Created    29s   kubelet, maysun-okd-worker3  Created container
  Normal  Started    29s   kubelet, maysun-okd-worker3  Started container
  Normal  Pulling    29s   kubelet, maysun-okd-worker3  pulling image "quay.io/eclipse/che-java8-maven:nightly"
  Normal  Pulled     28s   kubelet, maysun-okd-worker3  Successfully pulled image "quay.io/eclipse/che-java8-maven:nightly"
  Normal  Created    28s   kubelet, maysun-okd-worker3  Created container
  Normal  Started    28s   kubelet, maysun-okd-worker3  Started container
```

Integration tests to be covered by #2673

kadel commented:
/approve

rajivnathan commented on pkg/devfile/adapters/kubernetes/utils/utils.go:
```diff
@@ -44,6 +44,21 @@ func GetContainers(devfileObj devfile.DevfileObj) []corev1.Container {
 }
 
+// GetVolumes iterates through the components in the devfile and returns a map of component alias to the devfile volumes
+func GetVolumes(devfileObj devfile.DevfileObj) map[string][]common.DockerimageVolume {
```
Should the return type be something more generic than `common.DockerimageVolume`? When we support Kubernetes or other component types might be weird for them to have to convert to use the Docker image the type. I'm on the fence about this though because I think if we do create a generic type it will likely have the same properties.

maysunfaisal replied:
updated

rajivnathan commented on pkg/kclient/volumes.go:
```diff
@@ -13,6 +15,9 @@ import (
 	PersistentVolumeClaimAPIVersion = "v1"
+
+	// The length of the string to be generated for names of resources
+	nameLength = 5
```
I'm assuming we'd want to use this constant through kclient so makes sense to move it to kclient.go?

maysunfaisal replied:
updated

rajivnathan commented on pkg/kclient/volumes.go:
```diff
@@ -43,16 +48,16 @@ func AddPVCToPodTemplateSpec(podTemplateSpec *corev1.PodTemplateSpec, pvc, volum
-func AddVolumeMountToPodTemplateSpec(podTemplateSpec *corev1.PodTemplateSpec, volumeName, pvc string, containerMountPathsMap map[string][]string) error {
+// AddVolumeMountToPodTemplateSpec adds the Volume Mounts in componentAliasToMountPaths to the podTemplateSpec containers for a given pvc and volumeName
+// componentAliasToMountPaths is a map of a container alias to an array of its Mount Paths
+func AddVolumeMountToPodTemplateSpec(podTemplateSpec *corev1.PodTemplateSpec, volumeName, pvc string, componentAliasToMountPaths map[string][]string) error {
```
`pvc` doesn't appear to be used in this function

maysunfaisal replied:
updated

maysunfaisal commented:
@rajivnathan PR has been updated with your feedback, thanks for the help with the interfaces!

Tests after feedback refactor commit
```
maysuns-mbp:adapters maysun$ pwd
/Users/maysun/go/src/github.com/openshift/odo/pkg/devfile/adapters

maysuns-mbp:adapters maysun$ go clean -testcache && go test ./... -v
=== RUN   TestNewPlatformAdapter
=== RUN   TestNewPlatformAdapter/get_platform_adapter
--- PASS: TestNewPlatformAdapter (0.00s)
    --- PASS: TestNewPlatformAdapter/get_platform_adapter (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/devfile/adapters	0.763s
?   	github.com/openshift/odo/pkg/devfile/adapters/common	[no test files]
?   	github.com/openshift/odo/pkg/devfile/adapters/kubernetes	[no test files]
=== RUN   TestComponentAdapter
=== RUN   TestComponentAdapter/Case:_Invalid_devfile
=== RUN   TestComponentAdapter/Case:_Valid_devfile
 •  Waiting for component to start  ...
 ✓  Waiting for component to start [159065ns]
--- PASS: TestComponentAdapter (0.00s)
    --- PASS: TestComponentAdapter/Case:_Invalid_devfile (0.00s)
    --- PASS: TestComponentAdapter/Case:_Valid_devfile (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/devfile/adapters/kubernetes/component	0.687s
=== RUN   TestStorageCreate
=== RUN   TestStorageCreate/storage_test
--- PASS: TestStorageCreate (0.00s)
    --- PASS: TestStorageCreate/storage_test (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/devfile/adapters/kubernetes/storage	0.637s
?   	github.com/openshift/odo/pkg/devfile/adapters/kubernetes/utils	[no test files]
```

```
maysuns-mbp:kclient maysun$ pwd
/Users/maysun/go/src/github.com/openshift/odo/pkg/kclient

maysuns-mbp:kclient maysun$ go clean -testcache && go test ./... -v
=== RUN   TestCreateDeployment
=== RUN   TestCreateDeployment/Case:_Valid_deployment_name
=== RUN   TestCreateDeployment/Case:_Invalid_deployment_name
--- PASS: TestCreateDeployment (0.00s)
    --- PASS: TestCreateDeployment/Case:_Valid_deployment_name (0.00s)
    --- PASS: TestCreateDeployment/Case:_Invalid_deployment_name (0.00s)
=== RUN   TestUpdateDeployment
=== RUN   TestUpdateDeployment/Case:_Valid_deployment_name
=== RUN   TestUpdateDeployment/Case:_Invalid_deployment_name
--- PASS: TestUpdateDeployment (0.00s)
    --- PASS: TestUpdateDeployment/Case:_Valid_deployment_name (0.00s)
    --- PASS: TestUpdateDeployment/Case:_Invalid_deployment_name (0.00s)
=== RUN   TestGenerateContainer
=== RUN   TestGenerateContainer/#00
=== RUN   TestGenerateContainer/container1
--- PASS: TestGenerateContainer (0.00s)
    --- PASS: TestGenerateContainer/#00 (0.00s)
    --- PASS: TestGenerateContainer/container1 (0.00s)
=== RUN   TestGeneratePodTemplateSpec
=== RUN   TestGeneratePodTemplateSpec/podSpecTest
--- PASS: TestGeneratePodTemplateSpec (0.00s)
    --- PASS: TestGeneratePodTemplateSpec/podSpecTest (0.00s)
=== RUN   TestGeneratePVCSpec
=== RUN   TestGeneratePVCSpec/1
=== RUN   TestGeneratePVCSpec/2
--- PASS: TestGeneratePVCSpec (0.00s)
    --- PASS: TestGeneratePVCSpec/1 (0.00s)
    --- PASS: TestGeneratePVCSpec/2 (0.00s)
=== RUN   TestWaitAndGetPod
=== RUN   TestWaitAndGetPod/phase:_running
 •  Waiting for component to start  ...
 ✓  Waiting for component to start [164875ns]
=== RUN   TestWaitAndGetPod/phase:_failed
 •  Waiting for component to start  ...
 ✗  Waiting for component to start [91273ns]
=== RUN   TestWaitAndGetPod/phase:_unknown
 •  Waiting for component to start  ...
 ✗  Waiting for component to start [97284ns]
--- PASS: TestWaitAndGetPod (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_running (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_failed (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_unknown (0.00s)
=== RUN   TestCreatePVC
=== RUN   TestCreatePVC/Case:_Valid_pvc_name
=== RUN   TestCreatePVC/Case:_Invalid_pvc_name
--- PASS: TestCreatePVC (0.00s)
    --- PASS: TestCreatePVC/Case:_Valid_pvc_name (0.00s)
    --- PASS: TestCreatePVC/Case:_Invalid_pvc_name (0.00s)
=== RUN   TestAddPVCToPodTemplateSpec
=== RUN   TestAddPVCToPodTemplateSpec/podSpecTest
--- PASS: TestAddPVCToPodTemplateSpec (0.00s)
    --- PASS: TestAddPVCToPodTemplateSpec/podSpecTest (0.00s)
=== RUN   TestAddVolumeMountToPodTemplateSpec
=== RUN   TestAddVolumeMountToPodTemplateSpec/podSpecTest
=== RUN   TestAddVolumeMountToPodTemplateSpec/podSpecTest#01
--- PASS: TestAddVolumeMountToPodTemplateSpec (0.00s)
    --- PASS: TestAddVolumeMountToPodTemplateSpec/podSpecTest (0.00s)
    --- PASS: TestAddVolumeMountToPodTemplateSpec/podSpecTest#01 (0.00s)
=== RUN   TestGetPVCsFromSelector
=== RUN   TestGetPVCsFromSelector/Case:_Valid_pvc_name
=== RUN   TestGetPVCsFromSelector/Case:_Wrong_Label_Selector
--- PASS: TestGetPVCsFromSelector (0.00s)
    --- PASS: TestGetPVCsFromSelector/Case:_Valid_pvc_name (0.00s)
    --- PASS: TestGetPVCsFromSelector/Case:_Wrong_Label_Selector (0.00s)
=== RUN   TestAddPVCAndVolumeMount
=== RUN   TestAddPVCAndVolumeMount/Case:_Valid_case
=== RUN   TestAddPVCAndVolumeMount/Case:_Error_case
--- PASS: TestAddPVCAndVolumeMount (0.00s)
    --- PASS: TestAddPVCAndVolumeMount/Case:_Valid_case (0.00s)
    --- PASS: TestAddPVCAndVolumeMount/Case:_Error_case (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/kclient	1.764s
```

```
maysuns-mbp:odo maysun$ make golint
golangci-lint run ./... --timeout 5m
```

maysunfaisal commented on pkg/devfile/adapters/common/interface.go:
```diff
@@ -1,6 +1,16 @@
 
-// ComponentAdapter defines the functions that platform-specific adapters must implement
+import (
+	corev1 "k8s.io/api/core/v1"
```
updated

maysunfaisal commented on pkg/devfile/adapters/common/interface.go:
```diff
@@ -1,6 +1,16 @@
+// ComponentAdapter defines the component functions that platform-specific adapters must implement
 type ComponentAdapter interface {
-	Start() error
+	Initialize() (*corev1.PodTemplateSpec, error)
```
updated

maysunfaisal commented on pkg/devfile/adapters/common/interface.go:
```diff
@@ -1,6 +1,16 @@
+
+// StorageAdapter defines the storage functions that platform-specific adapters must implement
+type StorageAdapter interface {
+	Start(*corev1.PodTemplateSpec) error
```
updated

maysunfaisal commented on pkg/devfile/adapters/kubernetes/adapter.go:
```diff
@@ -3,29 +3,43 @@ package kubernetes
 func (k Adapter) Start() error {
 
-	err := k.componentAdapter.Start()
+	podTemplateSpec, err := k.componentAdapter.Initialize()
```
updated

maysunfaisal commented on pkg/devfile/adapters/kubernetes/utils/utils.go:
```diff
@@ -44,6 +44,21 @@ func GetContainers(devfileObj devfile.DevfileObj) []corev1.Container {
+	componentAliasToVolumes := make(map[string][]common.DockerimageVolume)
+	// Only components with aliases are considered because without an alias commands cannot reference them
+	for _, comp := range devfileObj.Data.GetAliasedComponents() {
+		if comp.Type == common.DevfileComponentTypeDockerimage {
```
updated

maysunfaisal commented:
@rajivnathan I have removed the storage adapter test because it's Create just calls function `CreateComponentStorage`, which is in `utils.go` and I have unit tests for that in `unit_test.go` and I have also moved all the unit testing of storage's `adapter_test.go` to `volumes_test.go`. The new `volumes_test.go` was basically just a cut paste of the previous storage adapter tests. You can have a look at this in the prev commit e22cd68d9266c090022cfd57e18e508cc6b9b4d5

maysunfaisal commented on pkg/devfile/adapters/common/utils.go:
```diff
@@ -0,0 +1,21 @@
+	var components []common.DevfileComponent
+	// Only components with aliases are considered because without an alias commands cannot reference them
+	for _, comp := range data.GetAliasedComponents() {
+		if comp.Type == common.DevfileComponentTypeDockerimage {
```
updated

maysunfaisal commented on pkg/devfile/adapters/kubernetes/storage/adapter.go:
```diff
@@ -20,34 +17,20 @@ func New(adapterContext common.AdapterContext, client kclient.Client) Adapter {
-
-	// Add PVC and Volume Mounts to the podTemplateSpec
-	err = kclient.AddPVCAndVolumeMount(podTemplateSpec, volumeNameToPVC, componentAliasToVolumes)
+	a.VolumeNameToPVCName, err = CreateComponentStorage(&a.Client, volumes, componentName)
```
updated

johnmcollier commented on pkg/kclient/kclient.go:
```diff
@@ -68,3 +69,18 @@ func CreateObjectMeta(name, namespace string, labels, annotations map[string]str
+func GenerateOwnerReference(deployment *appsv1.Deployment) metav1.OwnerReference {
+
+	ownerReference := metav1.OwnerReference{
+		APIVersion: "extensions/v1beta1",
```
To add on to what Rajiv said and why we might not want to reference it directly: we should not be using or referencing Deployments in the `extensions/v1beta1` API group.

Starting with Kube 1.16, it’s been removed from the Kube API and cant be referenced.

johnmcollier commented on pkg/devfile/adapters/common/utils.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+// GetSupportedComponents iterates through the components in the devfile and returns a list of odo supported components
+func GetSupportedComponents(data versions.DevfileData) []common.DevfileComponent {
```
Could you add a unit test for GetSupportedComponents?

maysunfaisal replied:
makes sense, let me see

johnmcollier commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -42,26 +45,63 @@ func (a Adapter) Start() (err error) {
 
 	objectMeta := kclient.CreateObjectMeta(componentName, a.Client.Namespace, labels, nil)
```
We're probably unlikely to hit this issue with component names, **but** to be safe, we should probably trim and sanitize `componentName` with either `utils.DNS1123Name()` or `utils.NamespaceOpenShiftObject` before passing it into CreateObjectMeta.

maysunfaisal replied:
updated

johnmcollier commented on pkg/kclient/volumes.go:
```diff
@@ -82,3 +111,36 @@ func (c *Client) GetPVCsFromSelector(selector string) ([]corev1.PersistentVolume
+// generateVolumeNameFromPVC generates a random volume name based on the name
+// of the given PVC
+func generateVolumeNameFromPVC(pvc string) string {
+	return fmt.Sprintf("%v-%v-volume", pvc, util.GenerateRandomString(nameLength))
```
Nit: Any reason why we wouldn't just want to call utils.NamespacedOpenShiftObject here?

maysunfaisal replied:
I think volume name is not limited to 63 limit since its not a kube resource(need to confirm tho).

johnmcollier replied:
You're right, it might be 253

maysunfaisal replied:
actually, does not allow my vol name to be more than 63, updated

maysunfaisal commented on pkg/devfile/adapters/kubernetes/storage/interface.go:
```diff
@@ -0,0 +1,6 @@
+package storage
+
+// HelperAdapter defines functions that kubernetes storage adapters may implement
+type HelperAdapter interface {
```
being removed with the latest commit

maysunfaisal commented on pkg/devfile/adapters/kubernetes/storage/utils.go:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	randomChars := util.GenerateRandomString(4)
+	namespaceKubernetesObject, err := util.NamespaceOpenShiftObject(name, componentName)
```
wanted to, but a lot of code in odo1 is using it. Still change the function name?

johnmcollier replied:
If it's just a rename, it should be safe to do in my opinion. @kadel what are your thoughts?

maysunfaisal commented on pkg/devfile/adapters/kubernetes/storage/utils_test.go:
```diff
@@ -0,0 +1,83 @@
+	volNames := [...]string{"vol1", "vol2", "vol3"}
+	volSize := "5Gi"
+
+	tests := []struct {
```
let me see what I can do with these

maysunfaisal commented:
/retest

maysunfaisal commented:
@rajivnathan @johnmcollier I've pushed a new commit, pls review at your convenience, thx!

kadel commented:
> @rajivnathan @johnmcollier I've pushed a new commit, pls review at your convenience, thx!

@rajivnathan @johnmcollier, after you review this you can just do `/lgtm` on a separate line in a comment and this PR will get automaticaly merged.

rajivnathan commented:
Looks good to me now. @johnmcollier please add the `/lgtm` once you're ready
